### PR TITLE
fix issue of actions dag split

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- ActionsDAG::split can't make sure that "Execution of first then second parts on block is equivalent to execution of initial DAG."

### Documentation entry for user-facing changes

Continues from https://github.com/ClickHouse/ClickHouse/pull/60206

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 32ed2353f511d9e87e48cf20d807a6fcdf6d9334 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61458/32ed2353f511d9e87e48cf20d807a6fcdf6d9334/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by alexey-milovidov — APPROVED** [review]

**Comment by Algunenano** [comment]
It failed a fast test and it seems related (doesn't fail without the change) so it seems this required a more detailed investigation. Reverting the offending PR is hard since a) there have been multiple changes after it and b) multiple fuzzer issues (aka crashes) were closed by those changes, so simply going back to the old code isn't a good option.

File: tests/queries/0_stateless/03010_view_prewhere_in.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
+DROP TABLE IF EXISTS v;
+CREATE VIEW v (`date` UInt32,`value` UInt8) AS
+WITH
+    data AS (SELECT '' id LIMIT 0),
+    r AS (SELECT'' as id, 1::UInt8 as value)
+SELECT
+    now() as date,
```
Comment (Algunenano) [inline tests/queries/0_stateless/03010_view_prewhere_in.sql:7]:
```suggestion
    1::UInt32 as date,
```

To use correct types. In the current release it reproduces:
```
Code: 352. DB::Exception: Received from clickhouse-01:49000. DB::Exception: Block structure mismatch in (columns with identical name must have identical structure) stream: different columns:
date UInt32 UInt32(size = 0)
date UInt32 Const(size = 0, UInt32(size = 1)). (AMBIGUOUS_COLUMN_NAME)
```

**Comment by Algunenano** [comment]
@taiyang-li We are changing the code to cover the broken test in a simpler way. Do you have some other test that might fail in this case? Note that the original patch in the PR introduced multiple columns as input which lead to them being loaded unnecessarily, which is why `02263_lazy_mark_load` failed.

**Comment by taiyang-li** [comment]
> @taiyang-li We are changing the code to cover the broken test in a simpler way. Do you have some other test that might fail in this case? Note that the original patch in the PR introduced multiple columns as input which lead to them being loaded unnecessarily, which is why `02263_lazy_mark_load` failed.

It seems still can't solve the issue described here: https://github.com/ClickHouse/ClickHouse/pull/60206#issuecomment-1955776208

**Comment by KochetovNicolai** [comment]
Integration tests (tsan) [4/6]  - #59631

**Comment by KochetovNicolai** [comment]
@taiyang-li in the example from https://github.com/ClickHouse/ClickHouse/pull/60206#issuecomment-1955776208 split is correct. 
The main invariant of `split` is that the execution of splitted DAGs one by one is equivalent of execution of original DAG. This is still true.
The part `First DAG and initial DAG have equal inputs, second DAG and initial DAG has equal outputs.` is not true anymore, I will remove the comment.

**Pull request merged**